crates/agglayer-errors/src/lib.rs

```diff
@@ -0,0 +1,469 @@
+
+    fn log_debug_with<F>(self, msg: F) -> Self
+    where
+        F: FnOnce() -> String,
+    {
+        if let Err(error) = &self {
+            tracing::debug!(?error, "{}", msg());
+        }
+        self
+    }
+
+    fn log_trace(self, msg: &str) -> Self {
+        if let Err(error) = &self {
+            tracing::trace!(?error, "{msg}");
+        }
+        self
+    }
+
+    fn log_trace_with<F>(self, msg: F) -> Self
+    where
+        F: FnOnce() -> String,
+    {
+        if let Err(error) = &self {
+            tracing::trace!(?error, "{}", msg());
+        }
+        self
+    }
+}
+
+/// Match an `eyre::Report` against multiple error types and patterns.
+///
+/// # Example
+/// ```rust
+/// use agglayer_errors::match_err;
+/// use eyre::Context as _;
+///
+/// #[derive(Debug, thiserror::Error)]
+/// #[error("Foo")]
+/// struct Foo;
+///
+/// #[derive(Debug, thiserror::Error)]
+/// #[error("Bar")]
+/// struct Bar(&'static str);
+///
+/// let err: eyre::Report = Foo.into();
+/// let s = match_err!(err,
+///     Foo: _ => "foo".to_string(),
+///     Bar: Bar("foo") => "bar/foo".to_string(),
+///     Bar: Bar(s) => format!("bar({s})"),
+///     @default: _ => "other".to_string(),
+/// );
+/// ```
+#[macro_export]
+macro_rules! match_err {
+    (
+        $expr:expr
+        $(, $ty:ty : $pat:pat => $handler:expr)*
+        $(, @default : $default_pat:pat => $default_handler:expr)?
+        $(,)?
+    ) => {
+        'result: {
+            let expr: $crate::eyre::Report = $expr;
+            $(
+                #[allow(unused_variables)] // Pattern variables are unused from this downcast_ref
+                if let Some($pat) = expr.downcast_ref::<$ty>() {
+                    let Ok($pat) = expr.downcast::<$ty>() else {
+                        // This should never happen because we just checked downcast_ref
+                        unreachable!("downcast_ref succeeded but downcast failed");
+                    };
+                    let res = $handler;
+                    break 'result res;
+                }
+            )*
+            $(
+                let $default_pat = expr;
+                $default_handler
+            )?
+        }
+    };
+}
+
+/// Match an `eyre::Result` against multiple `Ok` patterns and multiple error
+/// types and patterns.
+///
+/// If no error pattern matches, propagate the error up.
+///
+/// # Example
+/// ```rust
+/// use agglayer_errors::match_res;
+/// use eyre::Context as _;
+///
+/// #[derive(Debug, thiserror::Error)]
+/// #[error("Foo")]
+/// struct Foo;
+///
+/// #[derive(Debug, thiserror::Error)]
+/// #[error("Bar")]
+/// struct Bar(&'static str);
+///
+/// fn main() -> eyre::Result<()> {
+///     let res: eyre::Result<u32> = Err(Bar("foo").into());
+///     let s = match_res!(res,
+///         Ok: 42 => "the answer".to_string(),
+///         Ok: n => format!("ok({n})"),
+///         Err Bar: Bar("foo") => "bar/foo".to_string(),
+///         Err Bar: b => format!("bar({})", b.0),
```

Comment:
Could the syntax be changed so it resembles a bit more the usual pattern matching and type annotation syntax?
```suggestion
///         Ok(42) => "the answer".to_string(),
///         Ok(n) => format!("ok({n})"),
///         Err(Bar("foo"): Bar) => "bar/foo".to_string(),
///         Err(b: Bar) => format!("bar({})", b.0),
```

Reply:
Unfortunately no, I also had the same thought at first :sweat_smile: 
```
error: `$pat:pat` is followed by `:`, which is not allowed for `pat` fragments
   --> crates/agglayer-errors/src/lib.rs:152:22
    |
152 |         $(, $pat:pat : $ty:ty => $handler:expr)*
    |                      ^ not allowed after `pat` fragments
    |
    = note: allowed there are: `=>`, `,`, `=`, `if` or `in`
```

I guess it's already reserved syntax within `:pat` already.

I could go with `Bar("foo") if Bar`, but that'd look weird to me, in addition to potentially colliding with future extensions of the macro to allow `if` clauses.

It's actually the reason why I picked a completely unrelated syntax: this way at least we won't mistakenly take one for the other, it's a new syntax that has little relationship with the usual match syntax.

WDYT?

Reply:
Your syntax looks good! Unfortunately I wanted to use `type<_>(foo)` for the default branch, but `_` parses as a valid `:ty` fragment, so I had to settle with `else(foo)` for the "catch-all" case :smiling_face_with_tear: 

WDYT about the new syntax?